Writeup of error metrics and and arc approximation

raphlinus opened the pull request:
Writeup of error metrics and and arc approximation

This gets a bit more of my thinking down on paper about error metrics, and also presents the most important results for approximation by arcs. There are quite a few citations, the references are expanded.

It also starts a future work section.

The PR also contains a commit of some code for experimenting with arcs and error metrics. Included is a pretty good error metric for computing the approximate Fréchet distance between ESPC and arc, used to curve-fit the metric in the paper. Again that should be cleaned up and explained.

raphlinus committed:
Some experiments with arc fitting

There's more to be done, but checking in what I have. That includes very careful calculation of the area of an arc segment, and inverting that.

Also some explorations of area of an Euler spiral parallel curve. The result of that exploration is that fine-tuning area is not profitable, as it scales the same as the error of the underlying Euler spiral to arc approximation. A far better approach is to apply a small modifier to the error metric. The current `arc_main` contains a by-hand tester to fit that metric.

raphlinus committed:
Writeup of error metrics and and arc approximation

This gets a bit more of my thinking down on paper about error metrics, and also presents the most important results for approximation by arcs. There are quite a few citations, the references are expanded.

It also starts a future work section.

raphlinus committed:
Add arc_chord figure

raphlinus committed:
Start testbed stroke implementation

This copies much of the kurbo stroke implementation but adapts it for lowering. There are some improvements to the f64 flatten implementation, but it doesn't yet reflect all the robustness in the f32 version.

Both lowering to lines and arcs is supported.

raphlinus committed:
Wire up command line arguments to tests

This lets us stroke-expand the paths in SVG files. Current version is very minimal.

raphlinus committed:
Wire up skia's stroke expander

Use the skia-safe crate to be able to call into Skia. There's a bunch of rough edges, but it's enough to start getting comparisons.

armansito commented on paper.tex:
```diff
@@ -28,7 +28,7 @@ \section{Introduction}
 A number of factors contribute to an algorithm being ``GPU-friendly.'' In addition to simply being able to process the input segments in parallel, such an algorithm also avoids divergent control flow (avoiding the need for explicit subdivision at special events such as inflection points and cusps) and uses robust numerical techniques not subject to particular problems when evaluated using 32-bit  ...
 
-We propose that the correctness of stroke outlines be divided into \emph{weak correctness} and \emph{strong correctness.} We define strong correctness as the computation of the outline of a line swept along the segment, maintaining normal orientation, combined with stroke caps and joins. Weak correctness, by contrast, only requires the parallel curves of stroke segments, combined with caps and th ...
+We propose that the correctness of stroke outlines be divided into \emph{weak correctness} and \emph{strong correctness.} We define strong correctness as the computation of the outline of a line swept along the segment, maintaining normal orientation, combined with stroke caps and joins. Weak correctness, by contrast, only requires the parallel curves of stroke segments, combined with caps and th ...
```
> [TODO: find language in SVG standard]

[SVG 2, Section 13.5.7 Computing the Shape of a Stroke](https://www.w3.org/TR/SVG2/painting.html#StrokeShape) looks highly relevant.

armansito commented on paper.tex:
```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+% Note: this is being written as a fairly stand-alone section. No doubt some reorganization should happen
+
+The core problem in stroke expansion is approximating the desired curve by segments of some other curve, usually a simpler one. These segments must be within an error tolerance of the source curve, and ideally there are not many more of them than optimal. We consider a number of source to target pairs, most importantly cubic Béziers to Euler spirals, and Euler spiral parallel curves to either lin ...
```
> and ideally there are not many more of them than optimal.

Does it make sense to rephrase this to expand a bit on optiomal, for example:
_and ideally not exceed an amount that is optimal for rendering performance._

raphlinus replied:
I changed it to "and ideally close to a minimal number of them." Not exceeding sounds like a bound, and we're not claiming that. I'm open to revising further, we can flag this.

armansito commented on paper.tex:
```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+The core problem in stroke expansion is approximating the desired curve by segments of some other curve, usually a simpler one. These segments must be within an error tolerance of the source curve, and ideally there are not many more of them than optimal. We consider a number of source to target pairs, most importantly cubic Béziers to Euler spirals, and Euler spiral parallel curves to either lin ...
+
+There are generally three approaches to such curve approximation. The most straightforward but also least efficient is ``cut then measure,'' usually combined with adaptive subdivision. In this technique, a candidate approximate curve is produced, then the error is measured against the source curve, usually by sampling a number of points along both curves and determining a maximum error (or perhap ...
```
> (usually at t = 0.5)

Format `t = 0.5` as inline math?

armansito commented on paper.tex:
```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+The next approach is similar, but uses an \emph{error metric} to estimate the error. Ideally such a metric is a closed-form computation rather than requiring iteration. A good error metric is conservative, yet tight, in that it never underestimates the error (which would allow results exceeding the error bound to slip through), and does not significantly overestimate the error, which would result ...
+
+By far the most efficient approach is an \emph{invertible} error metric. In this approach, the error metric has an analytic inverse, or at least a good numerical approximation. Because the metric is invertible, it can predict the number of subdivisions needed, as well as the parameter value for each subdivision. If the error metric is accurate, then approximation is near-optimal. One example of a ...
```
> ...among other applications, it is used in Skia for path flattening. One limitation of Wang's method is that it only applies to the flattening of filled outlines, not parallel curves as is required for strokes.

There is potentially more to be said here. My first thought is about citation, since we're mentioning Skia, should we reference code? (I know where to point to here if necessary.)

My second thought is about the fact that Skia actually uses Wang's formula for tessellated GPU strokes using an algorithm that's very much akin to polar stroking. Graphite/Ganesh splits a given cubic into convex sub-cubics (aka "chopping") by analyzing cusps, 180 turns, and inflection points on the CPU. Each chopped segment gets encoded as a separate instance in an instanced triangle-strip draw (alongside joins and caps, which end up in their own instances). The vertex counts for any given instance is such that **a)** they are no less than the lower bound returned by Wang's formula and **b)** there are always enough vertices to tessellate a 180&deg; radial sweep (and then some) within the given tolerances. This use of Wang's formula is suitable for polar stroking in general, as the Kilgard'2020 states about their method:

> The number of tessellated quads must also be determined _a priori_ to tessellation of a given path segment, as opposed to being the result of a recursive process; this is motivated by wanting to map well to GPUs where predictable work creation is necessary as GPUs do not naturally support recursive processes.

I wanted to point this out because this scheme can draw strokes correctly (in the strongly correct sense) without computing a precise parallel curve which our text states as _required_. The sentence is not wrong per se but I think there is some nuance wrt the utility of Wang's formula and stroke rendering.[^1]

[^1]: This method of course has its down sides: the CPU analysis isn't cheap and the instance vertex counts can be far from what we would call "optimal" (many triangles become degenerate and are left unused), though there isn't a strong indication that the latter point (the suboptimal-ness of the subdivision count) has a significant impact on GPU performance. I should collect data on this.

raphlinus replied:
I've addressed this in the text. I think there are a bunch of different things going on here:

* In a local method, you can totally win by outputting a combination a combination of lines flattened by naive Wang and pie shapes between them; these pies are the same as joins (and it doesn't matter whether you generate them by further flattening to tris, or, for example, a fragment shader)

* You can fix Wang by doing a hybrid of Wang and polar stroking. That appears to be basically what Skia and Rive do. It represents state of the art, but is harder to talk about and benchmark, because it's not published as a citeable paper. A good case can be made that we should measure quantitatively, including the number of triangles generated, or possibly the number of outer contour edges.

armansito reviewed (commented):
I went over and commented on the first half of the text. I'll work through the error metric approximations tomorrow.

raphlinus committed:
Evolutes

This is a working implementation of evolutes. There are some limitations. The lowering of evolutes is crude. Also, there isn't an option to enable or disable them.

raphlinus committed:
Smart flattening of evolutes

Use fancy error metric to flatten evolutes to lines. Smart lowering to arcs still remains.

raphlinus committed:
Checkpoint ES evolute lowering to arcs

Fine-tuning the math

raphlinus committed:
Lower evolutes to arcs

The code now lowers Euler spiral segments to arcs, and it works well. The lowering logic is both in evolute.rs (lower_es_evolute_arc) and the Lowering implementation for arcs in stroke.rs

The PR also fixes a range bug in ESPC to arc lowering.

raphlinus committed:
Address review feedback in paper

This is a fairly small set of changes in response to review, including new insight into the hybrid nature of flattening in the Skia and Rive renderers.

raphlinus reviewed (commented):
I'm posting my responses to comments, and also pushing a minimal diff of changes in response to them. Likely we should land this (especially as it mixes code and text changes) and continue work in subsequent PRs.

armansito commented on paper.tex:
```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+\subsection{Error metrics for flattening}
+
+The distance between a circular arc segment of length $s$, with angle between arc and chord of $\theta$ (see Figure~\ref{fig:arc_chord}), is exactly $(1-\cos \theta)s/(2\theta)$. The curvature is $\kappa = \frac{2\theta}{s}$ (equivalently, $\theta = \frac{\kappa s}{2s}$), and this remains constant even as the arc is subdivided. Rewriting, $d = (1-\cos\frac{\kappa}{2s})\frac{s}{\kappa}$. From this ...
```
> ...$(1-\cos \theta)s/(2\theta)$...

nit: Use `\frac`? You used fractions elsewhere in the same paragraph and they look legible to me in the pdf.

$(1-\cos \theta)\frac{s}{2\theta}$

armansito replied:
> The distance between a circular arc segment of length $s$, with angle between arc and chord...

_The distance between a circular arc segment of length $s$_ **and its chord**, _with angle between arc and chord..._

armansito commented on paper.tex:
```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+\subsection{Error metrics for flattening}
+
+The distance between a circular arc segment of length $s$, with angle between arc and chord of $\theta$ (see Figure~\ref{fig:arc_chord}), is exactly $(1-\cos \theta)s/(2\theta)$. The curvature is $\kappa = \frac{2\theta}{s}$ (equivalently, $\theta = \frac{\kappa s}{2s}$), and this remains constant even as the arc is subdivided. Rewriting, $d = (1-\cos\frac{\kappa}{2s})\frac{s}{\kappa}$. From this ...
```
> (equivalently, $\theta = \frac{\kappa s}{2s}$),

Shouldn't this be $\theta = \frac{\kappa s}{2}$ ? Unless I'm mistaken, the next step in the derivation should then be:

  $$ d = (1-\cos\frac{\kappa s}{2})\frac{1}{\kappa} $$

Which then follows into the equation for $n$ segments as you have it, i.e:

  $$ d = (1-\cos\frac{\theta}{n})\frac{1}{\kappa} $$

  $$ d = (1-\cos\frac{\kappa s}{2n})\frac{1}{\kappa} $$

raphlinus replied:
Yes thanks. I reworked this section a bunch of times and didn't manage to entirely clean up the mess.

armansito commented on paper.tex:
```diff
@@ -83,7 +159,7 @@ \section{Flattened parallel curves}
 
```
Above:

> Analytically, is is a tenth order algebraic curve, whic

...**it** _is a tenth order_...

armansito commented on paper.tex:
```diff
@@ -169,6 +235,36 @@ \subsection{Approximation of the subdivision density integral}
 
+\section{Error metrics for approximation by arcs}
+
+The problem of approximating a curve by a sequence of arc segments has extensive literature, but none of the published solutions are quite suitable for our application. The specific problem of approximating an Euler spiral by arcs is considered in \cite{Meek2004} using a ``cut then measure'' adaptive subdivision scheme, but their solution has poor quality; it scales as $O(1/n^2)$, while $O(1/n^3) ...
```
> they are solving a harder problem, adopting the constraint that the generated arc sequence...

nit: Use `:` instead of comma after "they are solving a harder problem"?

armansito replied:
> While desirable for many applications, for simply rendering a stroked outline, it is not needed.

nit: slight rephrasing suggestion: ..._While desirable for many applications,_ **this constraint is not needed for rendering a stroke outline.**

armansito replied:
> discontinuities of an arc approximation are tiny compared to the flattening to lines.

nit: I would drop "the" in "the flattening to lines".

armansito reviewed (approved):
I carefully checked the math where I could and suggested some further edits. Thank you for the followups to the earlier review, I think the updated sections read a lot better now.

Let's land this now and plan for a later edit pass closer to the submission.

A note for later: should a definition of scaling come earlier in the paper? The section 4 draft currently introduces the term but it would be good to establish what it means before then.

raphlinus committed:
Address review feedback

Fix math in derivation of simple arc distance. A few phrasing changes.
